src/peergos/shared/hamt/Champ.java

```diff
@@ -0,0 +1,524 @@
+                        (a, b) -> a)
+                );
+    }
+
+    private Champ copyAndSetValue(final int setIndex, final int payloadIndex, final MaybeMultihash val) {
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] dst = new HashPrefixPayload[src.length];
+
+        System.arraycopy(src, 0, dst, 0, src.length);
+        HashPrefixPayload existing = dst[setIndex];
+        KeyElement[] updated = new KeyElement[existing.mappings.length];
+        System.arraycopy(existing.mappings, 0, updated, 0, existing.mappings.length);
+        updated[payloadIndex] = new KeyElement(existing.mappings[payloadIndex].key, val);
+        dst[setIndex] = new HashPrefixPayload(updated, null);
+
+        return new Champ(dataMap, nodeMap, dst);
+    }
+
+    private Champ insertIntoPrefix(final int index, final ByteArrayWrapper key, final MaybeMultihash val) {
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] result = new HashPrefixPayload[src.length];
+
+        System.arraycopy(src, 0, result, 0, src.length);
+        KeyElement[] prefix = new KeyElement[src[index].mappings.length + 1];
+        System.arraycopy(src[index].mappings, 0, prefix, 0, src[index].mappings.length);
+        prefix[prefix.length - 1] = new KeyElement(key, val);
+        // ensure canonical structure
+        Arrays.sort(prefix, Comparator.comparing(m -> m.key));
+        result[index] = new HashPrefixPayload(prefix, null);
+
+        return new Champ(dataMap, nodeMap, result);
+    }
+
+    private Champ addNewPrefix(final int bitpos, final ByteArrayWrapper key, final MaybeMultihash val) {
+        final int insertIndex = index(dataMap, bitpos);
+
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] result = new HashPrefixPayload[src.length + 1];
+
+        System.arraycopy(src, 0, result, 0, insertIndex);
+        System.arraycopy(src, insertIndex, result, insertIndex + 1, src.length - insertIndex);
+        result[insertIndex] = new HashPrefixPayload(new KeyElement[]{new KeyElement(key, val)}, null);
+
+        BitSet newDataMap = BitSet.valueOf(dataMap.toByteArray());
+        newDataMap.set(bitpos);
+        return new Champ(newDataMap, nodeMap, result);
+    }
+
+    private Champ copyAndMigrateFromInlineToNode(final int bitpos, final Pair<Champ, Multihash> node) {
+
+        final int oldIndex = index(dataMap, bitpos);
+        final int newIndex = this.contents.length - 1 - index(nodeMap, bitpos);
+
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] dst = new HashPrefixPayload[src.length];
+
+        // copy 'src' and remove 1 element at position oldIndex and insert 1 element at position newIndex
+        assert oldIndex <= newIndex;
+        System.arraycopy(src, 0, dst, 0, oldIndex);
+        System.arraycopy(src, oldIndex + 1, dst, oldIndex, newIndex - oldIndex);
```

Comment:
Is it possible for ``newIndex`` to be smaller than ``oldIndex``? If so this will fail.

Reply:
Nope, because the inline mappings are all before the child mappings